src/core/fs.rs

```diff
@@ -117,3 +121,246 @@ pub fn set_group(path: &Path, group: &str) -> Result<(), io::Error> {
 
     Ok(())
 }
+
+#[cfg(target_os = "macos")]
+fn get_errno() -> libc::c_int {
+    // SAFETY: always safe to call
+    unsafe { *libc::__error() }
+}
+
+#[cfg(not(target_os = "macos"))]
+fn get_errno() -> libc::c_int {
+    // SAFETY: always safe to call
+    unsafe { *libc::__errno_location() }
+}
+
+fn set_fd_nonblocking(fd: RawFd) -> Result<(), io::Error> {
+    // SAFETY: always safe to call
+    let flags = unsafe { libc::fcntl(fd, libc::F_GETFL, 0) };
+    if flags < 0 {
+        return Err(io::Error::last_os_error());
+    }
+
+    // SAFETY: always safe to call
+    let ret = unsafe { libc::fcntl(fd, libc::F_SETFL, flags | libc::O_NONBLOCK) };
+    if ret != 0 {
+        return Err(io::Error::last_os_error());
+    }
+
+    Ok(())
+}
+
+struct FileWatcherState {
+    watcher: notify::RecommendedWatcher,
+    changed: bool,
+}
+
+struct FileWatcherData {
+    file: PathBuf,
+    read_fd: RawFd,
+    write_fd: RawFd,
+    state: Mutex<Option<FileWatcherState>>,
+}
+
+pub struct FileWatcher {
+    data: Arc<FileWatcherData>,
+}
+
+#[derive(Debug)]
+pub struct FileWatcherError;
+
+impl FileWatcher {
+    pub fn new<P: AsRef<Path>>(file_path: P) -> Result<Self, FileWatcherError> {
+        let file = file_path.as_ref();
+
+        let dir = match file.parent() {
+            Some(p) => p,
+            None => return Err(FileWatcherError),
+        };
+
+        let mut fds = [0; 2];
+
+        // SAFETY: fds pointer is valid
+        let ret = unsafe { libc::pipe(fds.as_mut_ptr()) };
+        assert_eq!(ret, 0);
+
+        for fd in &fds {
+            set_fd_nonblocking(*fd).unwrap();
```

Comment:
We return a `FileWatcherError` here, why do we not bubble this unwrap up (along with a few others around obtaining locks) and panic here instead, should this fail if we cannot set `nonblocking` on _any_ of the files?

Reply:
As discussed, this operation shouldn't fail and isn't really recoverable anyway.